subsys/usb/class/netusb/function_rndis.c

```diff
@@ -0,0 +1,1070 @@
+NET_BUF_POOL_DEFINE(rndis_cmd_pool, CONFIG_RNDIS_CMD_BUF_COUNT,
+		    CONFIG_RNDIS_CMD_BUF_SIZE, 0, NULL);
+static struct k_fifo rndis_cmd_queue;
+
+static struct k_delayed_work notify_work;
+
+/*
+ * Stack for cmd thread
+ */
+static K_THREAD_STACK_DEFINE(cmd_stack, 2048);
+static struct k_thread cmd_thread_data;
+
+/*
+ * TLV structure is used for data encapsulation parsing
+ */
+struct tlv {
+	u32_t type;
+	u32_t len;
+	u8_t data[];
+} __packed;
+
+static struct __rndis {
+	u32_t net_filter;
+
+	enum {
+		UNINITIALIZED,
+		INITIALIZED,
+	} state;
+
+	struct net_pkt *in_pkt;	/* Pointer to pkt assembling at the moment */
+	int in_pkt_len;		/* Packet length to be assembled */
+	int skip_bytes;		/* In case of low memory, skip bytes */
+
+	u16_t mtu;
+	u16_t speed;		/* TODO: Calculate right speed */
+
+	/* Statistics */
+	u32_t rx_err;
+	u32_t tx_err;
+	u32_t rx_no_buf;
+
+	atomic_t notify_count;
+
+	u8_t mac[6];
+	u8_t media_status;
+} rndis = {
+	.mac =  { 0x00, 0x11, 0x22, 0x33, 0x44, 0x66 },
+	.mtu = 1500, /* Ethernet frame */
+	.media_status = RNDIS_OBJECT_ID_MEDIA_DISCONNECTED,
+	.state = UNINITIALIZED,
+	.skip_bytes = 0,
+	.speed = 0,
+};
+
+static u8_t manufacturer[] = CONFIG_USB_DEVICE_MANUFACTURER;
+static u32_t drv_version = 1;
+
+static u32_t object_id_supported[] = {
+	RNDIS_OBJECT_ID_GEN_SUPP_LIST,
+	RNDIS_OBJECT_ID_GEN_HW_STATUS,
+	RNDIS_OBJECT_ID_GEN_SUPP_MEDIA,
+	RNDIS_OBJECT_ID_GEN_IN_USE_MEDIA,
+
+	RNDIS_OBJECT_ID_GEN_MAX_FRAME_SIZE,
+	RNDIS_OBJECT_ID_GEN_LINK_SPEED,
+	RNDIS_OBJECT_ID_GEN_BLOCK_TX_SIZE,
+	RNDIS_OBJECT_ID_GEN_BLOCK_RX_SIZE,
+
+	RNDIS_OBJECT_ID_GEN_VENDOR_ID,
+	RNDIS_OBJECT_ID_GEN_VENDOR_DESC,
+	RNDIS_OBJECT_ID_GEN_VENDOR_DRV_VER,
+
+	RNDIS_OBJECT_ID_GEN_PKT_FILTER,
+	RNDIS_OBJECT_ID_GEN_MAX_TOTAL_SIZE,
+	RNDIS_OBJECT_ID_GEN_CONN_MEDIA_STATUS,
+	RNDIS_OBJECT_ID_GEN_PHYSICAL_MEDIUM,
+#if defined(USE_RNDIS_STATISTICS)
+	/* Using RNDIS statistics puts heavy load on
+	 * USB bus, disable it for now
+	 */
+	RNDIS_OBJECT_ID_GEN_TRANSMIT_OK,
+	RNDIS_OBJECT_ID_GEN_RECEIVE_OK,
+	RNDIS_OBJECT_ID_GEN_TRANSMIT_ERROR,
+	RNDIS_OBJECT_ID_GEN_RECEIVE_ERROR,
+	RNDIS_OBJECT_ID_GEN_RECEIVE_NO_BUF,
+#endif /* USE_RNDIS_STATISTICS */
+	RNDIS_OBJECT_ID_802_3_PERMANENT_ADDRESS,
+	RNDIS_OBJECT_ID_802_3_CURR_ADDRESS,
+	RNDIS_OBJECT_ID_802_3_MCAST_LIST,
+	RNDIS_OBJECT_ID_802_3_MAX_LIST_SIZE,
+	RNDIS_OBJECT_ID_802_3_MAC_OPTIONS,
+};
+
+static int parse_rndis_header(const u8_t *buffer, u32_t buf_len)
+{
+	struct rndis_payload_packet *hdr = (void *)buffer;
+	u32_t len;
+
+	if (buf_len < sizeof(*hdr)) {
+		SYS_LOG_ERR("Too small packet len %u", buf_len);
+		return -EINVAL;
+	}
+
+	if (hdr->type != sys_cpu_to_le32(RNDIS_DATA_PACKET)) {
+		SYS_LOG_ERR("Wrong data packet type 0x%x",
+			    sys_le32_to_cpu(hdr->type));
+		return -EINVAL;
+	}
+
+	len = sys_le32_to_cpu(hdr->len);
+	/*
+	 * Calculate additional offset since payload_offset is calculated
+	 * from the start of itself ;)
+	 */
+	if (len < sys_le32_to_cpu(hdr->payload_offset) +
```

Comment:
Would there be any issue if `payload_offset + payload_len - offsetof(header, payload_offset)` overflowed?  Unsigned overflow would mean that this would wrap around, creating a value that is smaller than `len`.

Reply:
We get from this function only len, of course USB device could put anything to any field and we cannot always verify it, so in a case packet is broken it would be dropped by network interface.

Reply:
@lpereira what do you think?